Remove cp932 dependency

trueroad opened the pull request:
https://github.com/texjporg/jfontmaps/issues/27#issuecomment-601068543
にて指摘させていただきましたが、Windows 環境では CP932 に依存した作りになっていて、他言語版 Windows の場合や、日本語版 Windows でも active codepage が CP65001 (UTF-8) になっている場合などが考慮されていないようです。
そこで、これらの環境でも非US-ASCIIのファイルを取り扱えるようにしてみました。また、 Windows か否かの判定を極力なくして共通の処理に変えたので、非 Windows 環境においても同様に非 US-ASCII のファイル名が使えるようになっています。

一部、絶対に US-ASCII だけだろうと思われる部分はそのままにしてあるところがあります。

なお、 Windows環境におけるシンボリックリンクなどの処理については
https://github.com/texjporg/jfontmaps/issues/27#issuecomment-601494166
の haranoaji-tlpost のような処理への改良をしておらず、従来どおりのままです。（本件だけでもそれなりの分量があってレビューいただくのも時間がかかると思いますし、変更等必要になったら手戻りがあるのも何なので、本件をマージしていただいたてから検討して別途 pull request にしようかと思っています。）

trueroad committed:
Move `use` functions inside `if` block to out of the block

Even if a `use` function is inside an `if` block,
it will execute regardless of the `if` condition.
A `use` functions expands into a BEGIN block,
so it will be executed first no matter where placed in the source.

This commit moves `use` functions inside `if` block to out of the block.
For the above reasons, there is no effective difference.

trueroad committed:
Replace `$^W = 1` with `use warnings`

trueroad committed:
Add standard input/output encoding settings

This commit makes cjk-gs-integrate.pl performsproper encoding conversion
between internal strings and standard input/output.

trueroad committed:
Add `@ARGV` and `$0` encoding settings

trueroad committed:
Add support for non-CP932 Windows

cjk-gs-integrate.pl could not handle non-US-ASCII filenames
on non-Japanese Windows and Japanese Windows with UTF-8 active code page.
`encode_utftocp` and `encode_cptoutf` only worked correctly
on Japanese Windows with active code page 932.

This commit removes them and adds more general encoding and decoding.
This allows cjk-gs-integrate.pl to work properly
in other languages and other active code pages (e.g. 65001 UTF-8).

trueroad committed:
Add reading database encoding settings

trueroad committed:
Add dumping database encoding settings

trueroad committed:
Add finding font absolute path encoding settings

trueroad committed:
Add filename encoding settings

trueroad committed:
Add kpsewhich output encoding settings

trueroad committed:
Add backquote execution encoding settings

trueroad committed:
Delete win32 encoding comment

Issues on win32 encoding have been resolved.

trueroad committed:
Add output files encoding settings

trueroad commented:
以下のようなことが可能になっています。

# Windows 環境

tlperl, kpsewhich, tlgs の active codepage が一致している必要があります。（普通は一致しているハズです。）

非 US-ASCII なファイル名のフォントとして
`フォント\幻ノにじみ明朝.otf`
`フォント\ロゴたいぷゴシック.otf`
`フォント\07やさしさゴシック.ttf`
を置いておきます。

テスト用のデータベースファイルとして以下の内容を UTF-8 で記載した `cjkgs-日本語ファイル名.dat` を置いておきます。

```cjkgs-日本語ファイル名.dat
# 日本語ファイル名のテスト

Name: MaboroshinoNijimiMincho
Class: Japan
OTFname: 幻ノにじみ明朝.otf

Name: 07LogoTypeGothic7
Class: Japan
OTFname: ロゴたいぷゴシック.otf

Name: 07YasashisaGothic
Class: Japan
TTFname: 07やさしさゴシック.ttf
```

管理者権限で以下を実行します。

```
set OSFONTDIR=フォント
c:\texlive\2020\tlpkg\tlperl\bin\perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
```

日本語を含んだファイル名・フォルダ名でも、のデータベース読み込み、シンボリックリンク、出力フォルダ名、いずれも成功します。
続けて

```
c:\texlive\2020\tlpkg\tlperl\bin\perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット --remove
c:\texlive\2020\tlpkg\tlperl\bin\perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
c:\texlive\2020\tlpkg\tlperl\bin\perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
c:\texlive\2020\tlpkg\tlperl\bin\perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット --cleanup
```

のように、 remove、2 回連続インストール、cleanup、いずれも成功します。

# 非 Windows 環境

perl, kpsewhich, gs のロケールが一致している必要があります。（普通は一致しているハズです。）

非 US-ASCII なファイル名のフォントとして
`フォント/幻ノにじみ明朝.otf`
`フォント/ロゴたいぷゴシック.otf`
`フォント/07やさしさゴシック.ttf`
を置いておきます。

同じくテスト用のデータベースファイル `cjkgs-日本語ファイル名.dat` を置いておきます。

以下を実行します。

```
rm -fr アウトプット
OSFONTDIR=フォント perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
```

日本語を含んだファイル名・フォルダ名でも、のデータベース読み込み、シンボリックリンク、出力フォルダ名、いずれも成功します。
続けて

```
OSFONTDIR=フォント perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット --remove
OSFONTDIR=フォント perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
OSFONTDIR=フォント perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット
OSFONTDIR=フォント perl cjk-gs-integrate.pl --fontdef-add cjkgs-日本語ファイル名.dat -o アウトプット -cleanup
```

以上、 remove 、2 回連続インストール、 cleanup いずれも成功します。

aminophen commented:
ありがとうございます。まだ動かしていませんが，一件コメントだけ。

* Karl の r54419 のコミットメッセージで Encode::Locale が evidently not part of standard Perl だという趣旨のことが書かれています。

> 非 Windows 環境においても同様に非 US-ASCII のファイル名が使えるようになっています。

については，現在でも非 US-ASCII のファイル名を扱えているつもりでした。（例：macOS のヒラギノフォント）。非 Windows では殆ど UTF-8 が既定になっているはずなので，一切の変換は行なっていませんでしたが，まずかったでしょうか？

trueroad commented:
> Karl の r54419 のコミットメッセージで Encode::Locale が evidently not part of standard Perl だという趣旨のことが書かれています。

手元の環境だと入っていたので、使ってもいいかなぁと思っていました。有れば使うが無ければ代替するという方法も考えてみます。ただ、ほとんど同じものを実装しても意味がないので、どこまでやるか、ですが…。

> 非 Windows では殆ど UTF-8 が既定になっているはずなので，一切の変換は行なっていませんでしたが，まずかったでしょうか？

perl は内部と外部を明確に区別していて、たとえ外部が UTF-8 だったとしても、内部に入れる際には `decode('UTF-8', $foo)` などして内部文字列に変換しなければならないし、内部文字列を外部へ出す際には `encode('UTF-8', $foo)` などして外部向けに変換しなければならない、という認識です。
標準入出力については e738a74c6599bc494d357eaff815e1d762d4103e の `binmode` でこの変換をするよう指定しているので、`print` などでは明示的に変換しなくても変換してくれています。

US-ASCII だけなら変換しなくても大丈夫ですが、非US-ASCIIが混入する可能性があるのであれば変換しておいた方が無難と思います。

aminophen commented:
> perl は内部と外部を明確に区別していて、…変換しなければならない

あまり言語に詳しくないのでこの辺は知りませんでした。

> 有れば使うが無ければ代替する
> ほとんど同じものを実装しても意味がないので、どこまでやるか、ですが…。

有れば使うが無ければ“変換しない”が実装可能なら，それで十分と思います。現行版でも（原理はともかく）挙動上は非 Windows で“無変換”なまま非 US-ASCII なファイルを扱えているので…。

trueroad commented:
> 有れば使うが無ければ“変換しない”が実装可能なら，それで十分と思います。

いえ、「変換しなければならない」ので、 Encode::Locale が無くても変換する必要があります。少なくとも `decode('UTF-8', $foo)` とか `encode('UTF-8', $foo)` とかは Encode::Locale が無くても動作するので、こういうもので変換する必要があります。

Encode::Locale の中身をみていると部分的な再実装は可能な感じです。同梱してしまうのが一番簡単そうでありますが…そういうわけにもいかないでしょうから、落としどころとしては、例えば Windows は Encode::Locale 必須、非 Windows はあれば使うけどなければ UTF-8 固定の変換にする、とかかなぁ、と思っています。

> 現行版でも（原理はともかく）挙動上は非 Windows で“無変換”なまま非 US-ASCII なファイルを扱えているので…。

UTF-8文字列として認識しておらず、ただのバイナリ列として扱っていて、入口と出口でたまたまバイナリ列の並びが一致するので問題ないように見えているだけだと思います。入口から出口までの中身の文字列処理の内容によっては破綻します。

trueroad committed:
Limit warning suppression range

trueroad committed:
Add support for environments without Encode::Locale

For Windows, Encode::Locale is required
because it can get active code page and determine the character encoding.

For non-Windows, if the system has Encode::Locale,
use it to determine the encoding.
If not, all encodings are assumed to be UTF-8.

trueroad commented:
> Windows は Encode::Locale 必須、非 Windows はあれば使うけどなければ UTF-8 固定の変換にする

を実装してみました。

新たに Encode::Alias を使っています（Encode::Locale が内部で使っていた）が、これは perl 標準品の認識です。
https://perldoc.perl.org/Encode/Alias.html
